src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -859,21 +871,65 @@ def update_agents(cmd, client, resource_group_name, cluster_name, https_proxy=""
+            if container_registry_repository is None:
+                container_registry_repository = helm_values.get('systemDefaultValues').get('image').get('repository')
+            if container_registry_anonymous_pull is not None:
+                container_registry_username = ""
+                container_registry_password = ""
+            elif container_registry_username is None and container_registry_password is None:
+                container_registry_username = helm_values.get('systemDefaultValues').get('image').get('username')
+                container_registry_password = helm_values.get('systemDefaultValues').get('image').get('password')
+            if container_registry_agent_version is None:
+                container_registry_agent_version = connected_cluster.agent_version
+
+    if auto_upgrade is None:
+        auto_upgrade = helm_values.get('systemDefaultValues').get('azureArcAgents').get('autoUpdate')
+        if not auto_upgrade:
+            auto_upgrade = "false"
+
     # Retrieving Helm chart OCI Artifact location
-    registry_path = os.getenv('HELMREGISTRY') if os.getenv('HELMREGISTRY') else utils.get_helm_registry(cmd, config_dp_endpoint, dp_endpoint_dogfood, release_train_dogfood)
+    if container_registry_repository:
+        if auto_upgrade != "false":
+            auto_upgrade_error = 'Disable auto upgrade when using custom registry'
+            telemetry.set_exception(exception=auto_upgrade_error, fault_type=consts.Custom_Registry_Disable_Auto_Upgrade_Fault_Type, summary=auto_upgrade_error)
+            raise ArgumentUsageError(auto_upgrade_error, recommendation='use --auto_upgrade false')
+        if container_registry_agent_version is None:  # true only when we are moving from MCR to custom registry and version is not provided
+            agent_version_error = 'Arc agent version must be provided when using custom container registry'
+            telemetry.set_exception(exception=agent_version_error, fault_type=consts.Custom_Registry_Agent_Version_Required_Fault_Type, summary=agent_version_error)
+            raise ArgumentUsageError(agent_version_error, recommendation='use --agent-version <version>')
+        registry_path = "{}/azurearck8s/azure-arc-k8sagents:{}".format(container_registry_repository, container_registry_agent_version)
+        # Send private registry information to telemetry
+        telemetry.add_extension_event('connectedk8s', {'Context.Default.AzureCLI.CustomContainerRegistry': container_registry_repository})
+    else:
+        registry_path = os.getenv('HELMREGISTRY') if os.getenv('HELMREGISTRY') else utils.get_helm_registry(cmd, config_dp_endpoint, dp_endpoint_dogfood, release_train_dogfood)
 
     reg_path_array = registry_path.split(':')
     agent_version = reg_path_array[1]
 
     # Set agent version in registry path
-    if connected_cluster.agent_version is not None:
+    if not isCustomRegistryEnabled and connected_cluster.agent_version is not None:
```

Comment:
One scenario which could cause issue here:
Let's say a connected cluster was onboarded using mcr(with agent version 0.1.1)
Now, the customer uses this cmd to move from mcr to pcr. In this case, he will have to pass the --pcr-agent-version. And let's say that this --pcr-agent-version is 0.1.2, which is different from current agent version (0.1.1). 
In such a case, we will still be looking for 0.1.1. in pcr due to above condition.

Reply:
good catch, updated isCustomRegistryEnabled to True, when moving from MCR to PCR for the first time as well.